ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java

```diff
@@ -171,7 +198,21 @@ public long writeTo(int index, WritableByteChannel channel, long relativeOffset,
     long startOffset = options.getOffset() + relativeOffset;
     long sizeToRead = Math.min(maxSize, options.getMessageInfo().getSize() - relativeOffset);
     logger.trace("Blob Message Read Set position {} count {}", startOffset, sizeToRead);
-    long written = options.getChannel().transferTo(startOffset, sizeToRead, channel);
+    long written = 0;
+    long bufStartOffset = relativeOffset - options.getPrefetchedDataRelativeOffset();
+    if (options.getPrefetchedDataRelativeOffset() != -1 && bufStartOffset < 0) {
+      logger.warn("Store prefetch: RelativeOffset out of boundary.", relativeOffset,
+          options.getPrefetchedDataRelativeOffset());
+    }
+    if (options.getPrefetchedDataRelativeOffset() != -1 && bufStartOffset >= 0
```

Comment:
I am not entirely sure why `bufStartOffset` check needs to be done here. Plus, if that check fails you would go to the else part and end up doing a `transferTo()` even if prefetch option was true, which is not really correct.

This method puts the burden on the caller to do the right thing, so I think the additional checks are unnecessary (we don't do any such checks for `transferTo()`). I think all you need to make sure is to find and maintain the right offset to start reading in the case of a prefetch.

Reply:
We didn't do that in transferTo is because transferTo can do file read at any offset, but socketChannel.write(prefetchedData) only has a small range data to use. 

Based on code logic, the caller is sending right `relativeOffset`, but I am afraid we missed something. 

As long as PROD doesn't has any warn, I am good to remove the extra check. How do you think so?

Reply:
I think as long as you are ensuring that you are prefetching the right number of bytes, you should go with the assumption here that the `writeTo()` will be called for exactly that (even in the `transferTo()` case, if the caller was not doing the right thing, we would be returning bad data anyway).